File: system-test/java/src/main/java/com/mycompany/myshop/testkit/driver/adapter/api/client/controllers/OrderController.java
package com.mycompany.myshop.testkit.driver.adapter.api.client.controllers;

import com.mycompany.myshop.testkit.driver.port.dtos.ViewOrderResponse;
import com.mycompany.myshop.testkit.driver.port.dtos.PlaceOrderRequest;
import com.mycompany.myshop.testkit.driver.port.dtos.PlaceOrderResponse;
import com.mycompany.myshop.testkit.driver.adapter.api.client.dtos.errors.ProblemDetailResponse;
import com.mycompany.myshop.testkit.driver.adapter.shared.client.http.JsonHttpClient;
import com.mycompany.myshop.testkit.common.Result;

public class OrderController {
    private static final String ENDPOINT = "/api/orders";

    private final JsonHttpClient<ProblemDetailResponse> httpClient;

    public OrderController(JsonHttpClient<ProblemDetailResponse> httpClient) {
        this.httpClient = httpClient;
    }

    public Result<PlaceOrderResponse, ProblemDetailResponse> placeOrder(PlaceOrderRequest request) {
        return httpClient.post(ENDPOINT, request, PlaceOrderResponse.class);
    }

    public Result<ViewOrderResponse, ProblemDetailResponse> viewOrder(String orderNumber) {
        return httpClient.get(ENDPOINT + "/" + orderNumber, ViewOrderResponse.class);
    }

    public Result<Void, ProblemDetailResponse> cancelOrder(String orderNumber) {
        return httpClient.post(ENDPOINT + "/" + orderNumber + "/cancel");
    }

    public Result<Void, ProblemDetailResponse> deliverOrder(String orderNumber) {
        return httpClient.post(ENDPOINT + "/" + orderNumber + "/deliver");
    }

}

File: system-test/dotnet/Driver.Adapter/Api/Client/Controllers/OrderController.cs
using Common;

using Driver.Adapter.Shared.Client.Http;

using Driver.Adapter.Api.Client.Dtos.Errors;

using Driver.Port.Dtos;



namespace Driver.Adapter.Api.Client.Controllers;



public class OrderController

{

    private const string Endpoint = "/api/orders";



    private readonly JsonHttpClient<ProblemDetailResponse> _httpClient;



    public OrderController(JsonHttpClient<ProblemDetailResponse> httpClient)

    {

        _httpClient = httpClient;

    }



    public Task<Result<PlaceOrderResponse, ProblemDetailResponse>> PlaceOrderAsync(PlaceOrderRequest request)

        => _httpClient.PostAsync<PlaceOrderResponse>(Endpoint, request);



    public Task<Result<ViewOrderResponse, ProblemDetailResponse>> ViewOrderAsync(string? orderNumber)

        => _httpClient.GetAsync<ViewOrderResponse>($"{Endpoint}/{orderNumber}");



    public Task<Result<VoidValue, ProblemDetailResponse>> CancelOrderAsync(string? orderNumber)
        => _httpClient.PostAsync($"{Endpoint}/{orderNumber}/cancel");

    public Task<Result<VoidValue, ProblemDetailResponse>> DeliverOrderAsync(string? orderNumber)
        => _httpClient.PostAsync($"{Endpoint}/{orderNumber}/deliver");

}









